[core-gl-kit] app-emulation/libnvidia-container Fix build issue with libnvidia-container on Go 1.24

org-tekeli-borisp opened the pull request:
Updated libnvidia-container to resolve build errors caused by Go 1.24 no longer supporting methods on C types. Introduced a patch converting the method to a standalone function and updated the package template to apply the patch. This ensures compatibility with Go 1.24 and successful builds.

(cherry picked from commit bb95958a3f144fe030597e8334c5bdfec40cf671) (cherry picked from commit cf61c08ca27a4bc25fa109bc366863973e28c3b8)

org-tekeli-borisp committed:
[core-gl-kit] app-emulation/libnvidia-container Fix build issue with libnvidia-container on Go 1.24

Updated libnvidia-container to resolve build errors caused by Go 1.24 no longer supporting methods on C types. Introduced a patch converting the method to a standalone function and updated the package template to apply the patch. This ensures compatibility with Go 1.24 and successful builds.

(cherry picked from commit bb95958a3f144fe030597e8334c5bdfec40cf671)
(cherry picked from commit cf61c08ca27a4bc25fa109bc366863973e28c3b8)

geaaru commented:
Have you tried to compile libnvidia-container with go 1.23? I'm not sure if it's need add a condition about this.

cuantar commented:
https://github.com/macaroni-os/kit-fixups/pull/308

... but, now that we have Go 1.24.1 in the tree, unslotted, don't we need this patch to avoid breaking on `emerge @world -NuD ...` ?

org-tekeli-borisp commented:
> #308
> 
> ... but, now that we have Go 1.24.1 in the tree, unslotted, don't we need this patch to avoid breaking on `emerge @world -NuD ...` ?

exactly

geaaru commented:
The question is: if someone needs go 1.23 that currently it's fine for the libnvidia-container version he can to compile the code with this patch always applied? I think not. It seems a breaking API change for a fast check. I never see that syntax, IMHO.
So, probably the patch could be applied in `post_src_unpack` only if go version is 1.24:

```
local gov=$(go version | awk '{ print $3 }' | sed -e 's/go//' -e 's/.[0-9]*$//')
if [ $gov -ge 1.24 ] ; then
   eapply ${FILESDIR}/xxx
fi
```

org-tekeli-borisp commented:
Hmnm, 

```
export CGO_CFLAGS="-std=gnu11 -O2"; \
export CGO_LDFLAGS="-Wl,--gc-sections -Wl,-s -Wl,-soname,libnvidia-container-go.so.1"; \
go build -o libnvidia-container-go.so -ldflags "-s -w" -buildmode=c-shared .
# nvcgo
./main.go:35:10: cannot define new methods on non-local type CDeviceRule
make[2]: *** [Makefile:40: libnvidia-container-go.so] Error 1
```

org-tekeli-borisp commented:
``` 
╰ $ go version | awk '{ print $3 }' | sed -e 's/go//' -e 's/.[0-9]*$//'
1.24
```

geaaru commented:
Please, check the `if` i wrote it just as an example

geaaru commented:
Ok, i have it. `-ge` permits only INTEGER. So:
```
local gov=$(go version | awk '{ print $3 }' | sed -e 's/go//' -e 's/.[0-9]*$//' -e 's/^[0-9].//')
if [ $gov -ge 24 ] ; then
 ...
```

org-tekeli-borisp commented:
Still the same
```
make[2]: Entering directory '/run/portage/app-emulation/libnvidia-container-1.17.5/work/libnvidia-container-1.17.5/deps/src/nvcgo'
rm -f -f libnvidia-container-go.so libnvidia-container-go.h
make[2]: Leaving directory '/run/portage/app-emulation/libnvidia-container-1.17.5/work/libnvidia-container-1.17.5/deps/src/nvcgo'
make -C /run/portage/app-emulation/libnvidia-container-1.17.5/work/libnvidia-container-1.17.5/deps/src/nvcgo VERSION=1.17.5 build
make[2]: Entering directory '/run/portage/app-emulation/libnvidia-container-1.17.5/work/libnvidia-container-1.17.5/deps/src/nvcgo'
export CGO_CFLAGS="-std=gnu11 -O2"; \
export CGO_LDFLAGS="-Wl,--gc-sections -Wl,-s -Wl,-soname,libnvidia-container-go.so.1"; \
go build -o libnvidia-container-go.so -ldflags "-s -w" -buildmode=c-shared .
# nvcgo
./main.go:35:10: cannot define new methods on non-local type CDeviceRule
make[2]: *** [Makefile:40: libnvidia-container-go.so] Error 1
make[2]: Leaving directory '/run/portage/app-emulation/libnvidia-container-1.17.5/work/libnvidia-container-1.17.5/deps/src/nvcgo'
make[1]: *** [/run/portage/app-emulation/libnvidia-container-1.17.5/work/libnvidia-container-1.17.5/mk/nvcgo.mk:31: build] Error 2
make[1]: Leaving directory '/run/portage/app-emulation/libnvidia-container-1.17.5/work/libnvidia-container-1.17.5'
make: *** [Makefile:259: deps] Error 2
 * ERROR: app-emulation/libnvidia-container-1.17.5::core-gl-kit-fixups failed (install phase):
 *   emake failed
 

```
![image](https://github.com/user-attachments/assets/81861700-bc76-4b93-8032-b9170d744940)

geaaru commented:
![image](https://github.com/user-attachments/assets/954fa180-e31e-48f5-913e-f806be9da8d5)
The if is not fixed yet ;)

org-tekeli-borisp commented:
if I change it to `24` insteed of `1.24` the next problem occurs 
```
 $ ebuild libnvidia-container-1.17.5.ebuild clean install
Appending /home/borisp/Development/src/macaronios/macaronios-kit-fixups/core-gl-kit/curated to PORTDIR_OVERLAY...
 * libnvidia-container-1.17.5-f23e5e5.tar.gz BLAKE2B SHA512 size ;-) ...                                                                                                                                              [ ok ]
 * nvidia-modprobe-550.54.14-149440c.tar.gz BLAKE2B SHA512 size ;-) ...                                                                                                                                               [ ok ]
 * elftoolchain-0.7.1.tar.bz2 BLAKE2B SHA512 size ;-) ...                                                                                                                                                             [ ok ]
>>> Unpacking source...
>>> Unpacking libnvidia-container-1.17.5-f23e5e5.tar.gz to /run/portage/app-emulation/libnvidia-container-1.17.5/work
>>> Unpacking nvidia-modprobe-550.54.14-149440c.tar.gz to /run/portage/app-emulation/libnvidia-container-1.17.5/work
>>> Unpacking elftoolchain-0.7.1.tar.bz2 to /run/portage/app-emulation/libnvidia-container-1.17.5/work
>>> Source unpacked in /run/portage/app-emulation/libnvidia-container-1.17.5/work
 * Applying libnvidia-container-1.17.4-go-1.24.patch ...
can't find file to patch at input line 22
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|https://patch-diff.githubusercontent.com/raw/NVIDIA/libnvidia-container/pull/297.patch
|
|From 1c680195fdc85948d635286b72a6ad9f823b5987 Mon Sep 17 00:00:00 2001
|From: Dmitry Sharshakov <dmitry.sharshakov@siderolabs.com>
|Date: Thu, 13 Feb 2025 10:18:59 +0100
|Subject: [PATCH] Fix building with Go 1.24
|
|Go 1.24 does not allow defining methods on C types anymore, so make convert a function, not a method.
|
|Fixes the following error when building with Go 1.24:
|`./main.go:35:10: cannot define new methods on non-local type CDeviceRule`
|
|Signed-off-by: Dmitry Sharshakov <dmitry.sharshakov@siderolabs.com>
|---
| src/nvcgo/main.go | 6 +++---
| 1 file changed, 3 insertions(+), 3 deletions(-)
|
|diff --git a/src/nvcgo/main.go b/src/nvcgo/main.go
|index 1523a06d..ed43be8e 100644
|--- a/src/nvcgo/main.go
|+++ b/src/nvcgo/main.go
--------------------------
No file to patch.  Skipping patch.
3 out of 3 hunks ignored                                                                                                                                                                                              [ !! ]
 * ERROR: app-emulation/libnvidia-container-1.17.5::core-gl-kit-fixups failed (unpack phase):
 *   patch -p1  failed with /run/portage/app-emulation/libnvidia-container-1.17.5/files/libnvidia-container-1.17.4-go-1.24.patch
 * 
 * Call stack:
 *               ebuild.sh, line   93:  Called post_src_unpack
 *             environment, line 1480:  Called eapply '/run/portage/app-emulation/libnvidia-container-1.17.5/files/libnvidia-container-1.17.4-go-1.24.patch'
 *             environment, line  621:  Called _eapply_patch '/run/portage/app-emulation/libnvidia-container-1.17.5/files/libnvidia-container-1.17.4-go-1.24.patch'
 *             environment, line  559:  Called __helpers_die 'patch -p1  failed with /run/portage/app-emulation/libnvidia-container-1.17.5/files/libnvidia-container-1.17.4-go-1.24.patch'
 *   isolated-functions.sh, line  112:  Called die
 * The specific snippet of code:
 *              die "$@"
 * 
 * If you need support, post the output of `emerge --info '=app-emulation/libnvidia-container-1.17.5::core-gl-kit-fixups'`,
 * the complete build log and the output of `emerge -pqv '=app-emulation/libnvidia-container-1.17.5::core-gl-kit-fixups'`.
 * The complete build log is located at '/run/portage/app-emulation/libnvidia-container-1.17.5/temp/build.log'.
 * The ebuild environment file is located at '/run/portage/app-emulation/libnvidia-container-1.17.5/temp/environment'.
 * Working directory: '/run/portage/app-emulation/libnvidia-container-1.17.5/work'
 * S: '/run/portage/app-emulation/libnvidia-container-1.17.5/work/libnvidia-container-1.17.5'

 ╭─ 🐺 borisp@blacksun  bash 5.2.37  🕑22:49:27  🖿 ~/Development/src/macaronios/macaronios-kit-fixups/core-gl-kit/curated/app-emulation/libnvidia-container 🧬[libnvidia-go-1.24] 📦 🐾7 
 ╰ $ doit
INFO     Cloning/updating kit-fixups to access generators (--fast to use as-is)                                                                                                                                              
INFO     executing: (cd /home/borisp/repo_tmp/source-trees/kit-fixups && git pull --ff-only)                                                                                                                                 
INFO     Autogen: app-emulation/libnvidia-container (latest)                                                                                                                                                                 
INFO     Created: libnvidia-container-1.17.5.ebuild                                                                                                                                                                          


 ╭─ 🐺 borisp@blacksun  bash 5.2.37  🕑22:49:51  🖿 ~/Development/src/macaronios/macaronios-kit-fixups/core-gl-kit/curated/app-emulation/libnvidia-container 🧬[libnvidia-go-1.24] 📦 🐾8 
 ╰ $ ebuild libnvidia-container-1.17.5.ebuild clean install
Appending /home/borisp/Development/src/macaronios/macaronios-kit-fixups/core-gl-kit/curated to PORTDIR_OVERLAY...
 * libnvidia-container-1.17.5-f23e5e5.tar.gz BLAKE2B SHA512 size ;-) ...                                                                                                                                              [ ok ]
 * nvidia-modprobe-550.54.14-149440c.tar.gz BLAKE2B SHA512 size ;-) ...                                                                                                                                               [ ok ]
 * elftoolchain-0.7.1.tar.bz2 BLAKE2B SHA512 size ;-) ...                                                                                                                                                             [ ok ]
>>> Unpacking source...
>>> Unpacking libnvidia-container-1.17.5-f23e5e5.tar.gz to /run/portage/app-emulation/libnvidia-container-1.17.5/work
>>> Unpacking nvidia-modprobe-550.54.14-149440c.tar.gz to /run/portage/app-emulation/libnvidia-container-1.17.5/work
>>> Unpacking elftoolchain-0.7.1.tar.bz2 to /run/portage/app-emulation/libnvidia-container-1.17.5/work
>>> Source unpacked in /run/portage/app-emulation/libnvidia-container-1.17.5/work
 * Applying libnvidia-container-1.17.4-go-1.24.patch ...
can't find file to patch at input line 22
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|https://patch-diff.githubusercontent.com/raw/NVIDIA/libnvidia-container/pull/297.patch
|
|From 1c680195fdc85948d635286b72a6ad9f823b5987 Mon Sep 17 00:00:00 2001
|From: Dmitry Sharshakov <dmitry.sharshakov@siderolabs.com>
|Date: Thu, 13 Feb 2025 10:18:59 +0100
|Subject: [PATCH] Fix building with Go 1.24
|
|Go 1.24 does not allow defining methods on C types anymore, so make convert a function, not a method.
|
|Fixes the following error when building with Go 1.24:
|`./main.go:35:10: cannot define new methods on non-local type CDeviceRule`
|
|Signed-off-by: Dmitry Sharshakov <dmitry.sharshakov@siderolabs.com>
|---
| src/nvcgo/main.go | 6 +++---
| 1 file changed, 3 insertions(+), 3 deletions(-)
|
|diff --git a/src/nvcgo/main.go b/src/nvcgo/main.go
|index 1523a06d..ed43be8e 100644
|--- a/src/nvcgo/main.go
|+++ b/src/nvcgo/main.go
--------------------------
No file to patch.  Skipping patch.
3 out of 3 hunks ignored                                                                                                                                                                                              [ !! ]
 * ERROR: app-emulation/libnvidia-container-1.17.5::core-gl-kit-fixups failed (unpack phase):
 *   patch -p1  failed with /run/portage/app-emulation/libnvidia-container-1.17.5/files/libnvidia-container-1.17.4-go-1.24.patch
 * 
 * Call stack:
 *               ebuild.sh, line   93:  Called post_src_unpack
 *             environment, line 1487:  Called eapply '/run/portage/app-emulation/libnvidia-container-1.17.5/files/libnvidia-container-1.17.4-go-1.24.patch'
 *             environment, line  621:  Called _eapply_patch '/run/portage/app-emulation/libnvidia-container-1.17.5/files/libnvidia-container-1.17.4-go-1.24.patch'
 *             environment, line  559:  Called __helpers_die 'patch -p1  failed with /run/portage/app-emulation/libnvidia-container-1.17.5/files/libnvidia-container-1.17.4-go-1.24.patch'
 *   isolated-functions.sh, line  112:  Called die
 * The specific snippet of code:
 *              die "$@"
 * 
 * If you need support, post the output of `emerge --info '=app-emulation/libnvidia-container-1.17.5::core-gl-kit-fixups'`,
```

geaaru commented:
Hold on, you need to apply patches after that the ${S} is been renamed.

org-tekeli-borisp commented:
> Hold on, you need to apply patches after that the ${S} is been renamed.

Tested it too

org-tekeli-borisp committed:
[core-gl-kit] app-emulation/libnvidia-container Add conditional patch application based on Go version

Previously, the patch was always applied regardless of the Go version. This change ensures the patch is only applied when the Go version is 1.24 or higher, improving compatibility and preventing unnecessary actions for older versions.

org-tekeli-borisp commented:
Done!
